crates/walrus-sui/src/client/retry_client/failover.rs

```diff
@@ -108,29 +112,36 @@ impl<ClientT, BuilderT: LazyClientBuilder<ClientT> + std::fmt::Debug>
         assert!(lazy_client_builders.len() >= max_tries);
 
         Ok(Self {
-            max_tries,
-            state: Arc::new(Mutex::new(FailoverState {
-                client: lazy_client_builders[0].lazy_build_client().await?,
-                rpc_url: lazy_client_builders[0].get_rpc_url().map(|s| s.to_string()),
-                current_index: 0,
-            })),
             lazy_client_builders,
+            state: Arc::new(Mutex::new(None)),
+            max_tries,
         })
     }
 
-    /// Returns the name of the current client.
-    pub async fn get_current_client(&self) -> Arc<ClientT> {
-        self.state.lock().await.client.clone()
+    /// Returns the current client.
+    pub async fn get_current_client(&self) -> Result<Arc<ClientT>, FailoverError> {
+        let mut state = self.state.lock().await;
+        match state.as_ref() {
+            Some(state) => Ok(state.client.clone()),
+            None => {
+                let client = self
+                    .fetch_next_client_inner(&mut BTreeSet::new(), &mut state)
+                    .await?;
+                Ok(client)
+            }
+        }
     }
 
-    /// Returns the name of the current client.
+    /// Returns the current client's RPC URL.
+    // This is a helper function is mainly used for logging.
     pub async fn get_current_rpc_url(&self) -> String {
```

Comment:
Now that the number of error cases for this routine has grown from 1 to 2, I think it is worth changing the return type to anyhow::Result\<String\>, and forcing the caller to understand that the returned value is not actually a URL. This would just bulletproof this method a bit more, so we don't have to read it and wonder whether the callers are trying to use this URL.

Reply:
Done.